File: tests/python/test_trendflex.py
"""
Python binding tests for TrendFlex indicator.
These tests mirror the Rust unit tests to ensure Python bindings work correctly.
"""
import pytest
import numpy as np
import sys
from pathlib import Path

try:
    import my_project as ta_indicators
except ImportError:

    try:
        import my_project as ta_indicators
    except ImportError:
        pytest.skip("Python module not built. Run 'maturin develop --features python' first", allow_module_level=True)

from test_utils import load_test_data, assert_close, EXPECTED_OUTPUTS
from rust_comparison import compare_with_rust


class TestTrendFlex:
    @pytest.fixture(scope='class')
    def test_data(self):
        return load_test_data()

    def test_trendflex_partial_params(self, test_data):
        """Test TrendFlex with partial parameters (None values) - mirrors check_trendflex_partial_params"""
        close = test_data['close']


        result = ta_indicators.trendflex(close)
        assert len(result) == len(close)

    def test_trendflex_accuracy(self, test_data):
        """Test TrendFlex matches expected values from Rust tests - mirrors check_trendflex_accuracy"""
        close = test_data['close']


        expected_last_five = [
            -0.19724678008015128,
            -0.1238001236481444,
            -0.10515389737087717,
            -0.1149541079904878,
            -0.16006869484450567,
        ]

        result = ta_indicators.trendflex(close, period=20)

        assert len(result) == len(close)


        assert_close(
            result[-5:],
            expected_last_five,
            rtol=1e-8,
            msg="TrendFlex last 5 values mismatch"
        )


        compare_with_rust('trendflex', result, 'close', {'period': 20})

    def test_trendflex_default_candles(self, test_data):
        """Test TrendFlex with default parameters - mirrors check_trendflex_default_candles"""
        close = test_data['close']


        result = ta_indicators.trendflex(close)
        assert len(result) == len(close)

    def test_trendflex_zero_period(self):
        """Test TrendFlex fails with zero period - mirrors check_trendflex_zero_period"""
        input_data = np.array([10.0, 20.0, 30.0])

        with pytest.raises(ValueError, match="period = 0|ZeroTrendFlexPeriod"):
            ta_indicators.trendflex(input_data, period=0)

    def test_trendflex_period_exceeds_length(self):
        """Test TrendFlex fails when period exceeds data length - mirrors check_trendflex_period_exceeds_length"""
        data_small = np.array([10.0, 20.0, 30.0])

        with pytest.raises(ValueError, match="period > data len|TrendFlexPeriodExceedsData"):
            ta_indicators.trendflex(data_small, period=10)

    def test_trendflex_very_small_dataset(self):
        """Test TrendFlex fails with insufficient data - mirrors check_trendflex_very_small_dataset"""
        single_point = np.array([42.0])

        with pytest.raises(ValueError, match="period > data len|TrendFlexPeriodExceedsData"):
            ta_indicators.trendflex(single_point, period=9)

    def test_trendflex_empty_input(self):
        """Test TrendFlex fails with empty input - mirrors check_trendflex_empty_input"""
        empty = np.array([])

        with pytest.raises(ValueError, match="No data provided|NoDataProvided"):
            ta_indicators.trendflex(empty)

    def test_trendflex_reinput(self, test_data):
        """Test TrendFlex applied twice (re-input) - mirrors check_trendflex_reinput"""
        close = test_data['close']


        first_result = ta_indicators.trendflex(close, period=20)
        assert len(first_result) == len(close)


        second_result = ta_indicators.trendflex(first_result, period=10)
        assert len(second_result) == len(first_result)


        if len(second_result) > 240:
            assert not np.any(np.isnan(second_result[240:])), "Found unexpected NaN after warmup period"

    def test_trendflex_nan_handling(self, test_data):
        """Test TrendFlex handles NaN values correctly - mirrors check_trendflex_nan_handling"""
        close = test_data['close']

        result = ta_indicators.trendflex(close, period=20)
        assert len(result) == len(close)


        if len(result) > 240:
            assert not np.any(np.isnan(result[240:])), "Found unexpected NaN after warmup period"


        first_valid = 0
        warmup = first_valid + 20


        assert np.all(np.isnan(result[:warmup])), f"Expected NaN in warmup period [0:{warmup})"

        assert not np.isnan(result[warmup]), f"Expected valid value at index {warmup}"

    def test_trendflex_streaming(self, test_data):
        """Test TrendFlex streaming matches batch calculation - mirrors check_trendflex_streaming"""
        close = test_data['close']
        period = 20


        batch_result = ta_indicators.trendflex(close, period=period)


        stream = ta_indicators.TrendFlexStream(period=period)
        stream_values = []

        for price in close:
            result = stream.update(price)
            stream_values.append(result if result is not None else np.nan)

        stream_values = np.array(stream_values)


        assert len(batch_result) == len(stream_values)


        for i, (b, s) in enumerate(zip(batch_result, stream_values)):
            if np.isnan(b) and np.isnan(s):
                continue
            assert_close(b, s, rtol=1e-9, atol=1e-9,
                        msg=f"TrendFlex streaming mismatch at index {i}")

    def test_trendflex_batch_single_param(self, test_data):
        """Test TrendFlex batch processing with single parameter - mirrors check_batch_default_row"""
        close = test_data['close']

        result = ta_indicators.trendflex_batch(
            close,
            period_range=(20, 20, 0),
        )

        assert 'values' in result
        assert 'periods' in result


        assert result['values'].shape[0] == 1
        assert result['values'].shape[1] == len(close)
        assert len(result['periods']) == 1
        assert result['periods'][0] == 20


        default_row = result['values'][0]
        expected_last_five = [
            -0.19724678008015128,
            -0.1238001236481444,
            -0.10515389737087717,
            -0.1149541079904878,
            -0.16006869484450567,
        ]


        assert_close(
            default_row[-5:],
            expected_last_five,
            rtol=1e-8,
            msg="TrendFlex batch default row mismatch"
        )


        single_result = ta_indicators.trendflex(close, period=20)
        assert_close(
            default_row,
            single_result,
            rtol=1e-10,
            msg="Batch vs single calculation mismatch"
        )

    def test_trendflex_batch_multiple_periods(self, test_data):
        """Test TrendFlex batch processing with multiple periods"""
        close = test_data['close'][:100]

        result = ta_indicators.trendflex_batch(
            close,
            period_range=(10, 30, 10),
        )

        assert 'values' in result
        assert 'periods' in result


        assert result['values'].shape[0] == 3
        assert result['values'].shape[1] == len(close)
        assert len(result['periods']) == 3
        assert list(result['periods']) == [10, 20, 30]


        for i, period in enumerate([10, 20, 30]):
            row_data = result['values'][i]
            single_result = ta_indicators.trendflex(close, period=period)
            assert_close(
                row_data,
                single_result,
                rtol=1e-10,
                msg=f"Batch row {i} (period={period}) mismatch"
            )


            warmup = period
            assert np.all(np.isnan(row_data[:warmup])), f"Expected NaN in warmup [0:{warmup}) for period={period}"
            assert not np.isnan(row_data[warmup]), f"Expected valid value at index {warmup} for period={period}"

    def test_trendflex_batch_edge_cases(self, test_data):
        """Test TrendFlex batch processing edge cases"""
        close = np.array([1.0, 2.0, 3.0, 4.0, 5.0, 6.0, 7.0, 8.0, 9.0, 10.0])


        single_batch = ta_indicators.trendflex_batch(
            close,
            period_range=(5, 5, 0),
        )
        assert single_batch['values'].shape[0] == 1
        assert single_batch['values'].shape[1] == len(close)
        assert len(single_batch['periods']) == 1


        large_step_batch = ta_indicators.trendflex_batch(
            close,
            period_range=(5, 7, 10),
        )

        assert large_step_batch['values'].shape[0] == 1
        assert large_step_batch['periods'][0] == 5


        with pytest.raises(ValueError, match="No data provided|All values are NaN"):
            ta_indicators.trendflex_batch(
                np.array([]),
                period_range=(20, 20, 0),
            )

    def test_trendflex_all_nan_input(self):
        """Test TrendFlex with all NaN values - mirrors check_trendflex_all_nan"""
        all_nan = np.full(100, np.nan)

        with pytest.raises(ValueError, match="All values are NaN|AllValuesNaN"):
            ta_indicators.trendflex(all_nan, period=20)

    def test_trendflex_invalid_period(self):
        """Test TrendFlex with various invalid period values"""
        data = np.array([1.0, 2.0, 3.0, 4.0, 5.0])


        with pytest.raises(ValueError, match="period = 0|ZeroTrendFlexPeriod"):
            ta_indicators.trendflex(data, period=0)


        with pytest.raises(ValueError, match="period > data len|TrendFlexPeriodExceedsData"):
            ta_indicators.trendflex(data, period=10)


        with pytest.raises(ValueError, match="period > data len|TrendFlexPeriodExceedsData"):
            ta_indicators.trendflex(data, period=len(data))

    def test_trendflex_warmup_calculation(self, test_data):
        """Test TrendFlex warmup period calculation"""

        test_periods = [5, 10, 20, 30, 50]
        close = test_data['close'][:200]

        for period in test_periods:
            if period >= len(close):
                continue

            result = ta_indicators.trendflex(close, period=period)


            first_valid = 0
            warmup = first_valid + period


            for i in range(warmup):
                assert np.isnan(result[i]), f"Expected NaN at index {i} for period={period}"


            if warmup < len(result):
                assert not np.isnan(result[warmup]), f"Expected valid value at index {warmup} for period={period}"

    def test_trendflex_batch_metadata(self, test_data):
        """Test TrendFlex batch metadata is correctly populated"""
        close = test_data['close'][:50]

        result = ta_indicators.trendflex_batch(
            close,
            period_range=(10, 20, 5),
        )


        assert 'periods' in result
        assert len(result['periods']) == 3
        assert list(result['periods']) == [10, 15, 20]


        assert result['values'].shape[0] == len(result['periods'])
        assert result['values'].shape[1] == len(close)


if __name__ == '__main__':
    pytest.main([__file__, '-v'])

File: tests/wasm/test_trendflex.js
/**
 * WASM binding tests for TrendFlex indicator.
 * These tests mirror the Rust unit tests to ensure WASM bindings work correctly.
 */
import test from 'node:test';
import assert from 'node:assert';
import path from 'path';
import { fileURLToPath } from 'url';
import { 
    loadTestData, 
    assertArrayClose, 
    assertClose,
    isNaN,
    assertAllNaN,
    assertNoNaN 
} from './test_utils.js';
import { compareWithRust } from './rust-comparison.js';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

let wasm;
let testData;


const EXPECTED_OUTPUTS = {
    trendflex: {
        defaultParams: { period: 20 },
        last5Values: [
            -0.19724678008015128,
            -0.1238001236481444,
            -0.10515389737087717,
            -0.1149541079904878,
            -0.16006869484450567,
        ]
    }
};

test.before(async () => {
    
    try {
        const wasmPath = path.join(__dirname, '../../pkg/my_project.js');
        const importPath = process.platform === 'win32' 
            ? 'file:///' + wasmPath.replace(/\\/g, '/')
            : wasmPath;
        wasm = await import(importPath);
    } catch (error) {
        console.error('Failed to load WASM module:', error);
        throw error;
    }
    
    
    testData = loadTestData();
});

test('trendflex_partial_params', () => {
    const close = testData.close;
    
    
    const result = wasm.trendflex_js(close, 20);
    assert.equal(result.length, close.length);
});

test('trendflex_accuracy', () => {
    const close = testData.close;
    const expected = EXPECTED_OUTPUTS.trendflex;
    
    const result = wasm.trendflex_js(close, expected.defaultParams.period);
    
    assert.equal(result.length, close.length);
    
    
    const last5 = result.slice(-5);
    assertArrayClose(
        last5,
        expected.last5Values,
        1e-8,
        "TrendFlex last 5 values mismatch"
    );
    
    
    compareWithRust('trendflex', result, 'close', expected.defaultParams);
});

test('trendflex_default_candles', () => {
    const close = testData.close;
    
    
    const result = wasm.trendflex_js(close, 20);
    assert.equal(result.length, close.length);
});

test('trendflex_zero_period', () => {
    const inputData = [10.0, 20.0, 30.0];
    
    assert.throws(() => {
        wasm.trendflex_js(inputData, 0);
    }, /period = 0|ZeroTrendFlexPeriod/);
});

test('trendflex_period_exceeds_length', () => {
    const dataSmall = [10.0, 20.0, 30.0];
    
    assert.throws(() => {
        wasm.trendflex_js(dataSmall, 10);
    }, /period > data len|TrendFlexPeriodExceedsData/);
});

test('trendflex_very_small_dataset', () => {
    const singlePoint = [42.0];
    
    assert.throws(() => {
        wasm.trendflex_js(singlePoint, 9);
    }, /period > data len|TrendFlexPeriodExceedsData/);
});

test('trendflex_empty_input', () => {
    const empty = [];
    
    assert.throws(() => {
        wasm.trendflex_js(empty, 20);
    }, /No data provided|NoDataProvided/);
});

test('trendflex_reinput', () => {
    const close = testData.close;
    
    
    const firstResult = wasm.trendflex_js(close, 20);
    assert.equal(firstResult.length, close.length);
    
    
    const secondResult = wasm.trendflex_js(firstResult, 10);
    assert.equal(secondResult.length, firstResult.length);
    
    
    if (secondResult.length > 240) {
        for (let i = 240; i < secondResult.length; i++) {
            assert.ok(!isNaN(secondResult[i]), `Unexpected NaN at index ${i}`);
        }
    }
});

test('trendflex_nan_handling', () => {
    const close = testData.close;
    
    const result = wasm.trendflex_js(close, 20);
    assert.equal(result.length, close.length);
    
    
    if (result.length > 240) {
        for (let i = 240; i < result.length; i++) {
            assert.ok(!isNaN(result[i]), `Unexpected NaN at index ${i}`);
        }
    }
    
    
    const firstValid = 0; 
    const warmup = firstValid + 20;
    
    
    assertAllNaN(result.slice(0, warmup), `Expected NaN in warmup period [0:${warmup})`);
    
    assert.ok(!isNaN(result[warmup]), `Expected valid value at index ${warmup}`);
});

test('trendflex_batch_single_param', () => {
    const close = testData.close;
    
    const result = wasm.trendflex_batch_js(close, 20, 20, 0);
    const metadata = wasm.trendflex_batch_metadata_js(20, 20, 0);
    
    
    assert.equal(metadata.length, 1);
    assert.equal(metadata[0], 20);
    
    
    assert.equal(result.length, close.length);
    
    
    const expected = EXPECTED_OUTPUTS.trendflex.last5Values;
    const last5 = result.slice(-5);
    assertArrayClose(
        last5,
        expected,
        1e-8,
        "TrendFlex batch last 5 values mismatch"
    );
    
    
    const singleResult = wasm.trendflex_js(close, 20);
    assertArrayClose(
        result,
        singleResult,
        1e-10,
        "Batch vs single calculation mismatch"
    );
});

test('trendflex_batch_multiple_periods', () => {
    const close = testData.close.slice(0, 100); 
    
    const result = wasm.trendflex_batch_js(close, 10, 30, 10);
    const metadata = wasm.trendflex_batch_metadata_js(10, 30, 10);
    
    
    assert.equal(metadata.length, 3);
    assert.deepEqual(Array.from(metadata), [10, 20, 30]);
    
    
    assert.equal(result.length, 3 * close.length);
    
    
    const periods = [10, 20, 30];
    for (let i = 0; i < periods.length; i++) {
        const rowStart = i * close.length;
        const rowEnd = rowStart + close.length;
        const rowData = result.slice(rowStart, rowEnd);
        
        const singleResult = wasm.trendflex_js(close, periods[i]);
        assertArrayClose(
            rowData,
            singleResult,
            1e-10,
            `Period ${periods[i]} mismatch`
        );
        
        
        const warmup = periods[i]; 
        assertAllNaN(rowData.slice(0, warmup), `Expected NaN in warmup [0:${warmup}) for period=${periods[i]}`);
        assert.ok(!isNaN(rowData[warmup]), `Expected valid value at index ${warmup} for period=${periods[i]}`);
    }
});

test('trendflex_all_nan_input', () => {
    const allNan = new Array(100).fill(NaN);
    
    assert.throws(() => {
        wasm.trendflex_js(allNan, 20);
    }, /All values are NaN|AllValuesNaN/);
});


test('trendflex_batch_ergonomic_single_parameter', () => {
    const close = testData.close;
    
    
    const batchResult = wasm.trendflex_batch(close, {
        period_range: [20, 20, 0]
    });
    
    
    const singleResult = wasm.trendflex_js(close, 20);
    
    assert.strictEqual(batchResult.values.length, singleResult.length);
    assertArrayClose(batchResult.values, singleResult, 1e-10, "Batch vs single mismatch");
});

test('trendflex_batch_ergonomic_multiple_periods', () => {
    const close = testData.close.slice(0, 100);
    
    
    const batchResult = wasm.trendflex_batch(close, {
        period_range: [10, 30, 10]      
    });
    
    
    assert.strictEqual(batchResult.values.length, 3 * 100);
    assert.strictEqual(batchResult.rows, 3);
    assert.strictEqual(batchResult.cols, 100);
    
    
    assert.strictEqual(batchResult.combos.length, 3);
    assert.strictEqual(batchResult.combos[0].period, 10);
    assert.strictEqual(batchResult.combos[1].period, 20);
    assert.strictEqual(batchResult.combos[2].period, 30);
});

test('trendflex_batch_edge_cases', () => {
    const close = new Float64Array(10);
    close.fill(100);
    
    
    const singleBatch = wasm.trendflex_batch(close, {
        period_range: [5, 5, 1]
    });
    
    assert.strictEqual(singleBatch.values.length, 10);
    assert.strictEqual(singleBatch.combos.length, 1);
    
    
    const largeBatch = wasm.trendflex_batch(close, {
        period_range: [5, 7, 10] 
    });
    
    
    assert.strictEqual(largeBatch.values.length, 10);
    assert.strictEqual(largeBatch.combos.length, 1);
    
    
    assert.throws(() => {
        wasm.trendflex_batch(new Float64Array([]), {
            period_range: [20, 20, 0]
        });
    }, /All values are NaN/);
});


test('trendflex_zero_copy_basic', () => {
    
    const data = new Float64Array(testData.close.slice(0, 100));
    const period = 20;
    
    
    const ptr = wasm.trendflex_alloc(data.length);
    assert(ptr !== 0, 'Failed to allocate memory');
    
    
    const memView = new Float64Array(
        wasm.__wasm.memory.buffer,
        ptr,
        data.length
    );
    
    
    memView.set(data);
    
    
    try {
        wasm.trendflex_into(ptr, ptr, data.length, period);
        
        
        
        const updatedMemView = new Float64Array(wasm.__wasm.memory.buffer, ptr, data.length);
        
        
        const regularResult = wasm.trendflex_js(data, period);
        for (let i = 0; i < data.length; i++) {
            
            const memValue = updatedMemView[i];
            const regValue = regularResult[i];
            
            if ((isNaN(regValue) || regValue === undefined) && 
                (isNaN(memValue) || memValue === undefined)) {
                continue; 
            }
            
            assert(Math.abs(regValue - memValue) < 1e-10,
                   `Zero-copy mismatch at index ${i}: regular=${regValue}, zerocopy=${memValue}`);
        }
    } finally {
        
        wasm.trendflex_free(ptr, data.length);
    }
});

test('trendflex_zero_copy_error_handling', () => {
    
    assert.throws(() => {
        wasm.trendflex_into(0, 0, 10, 20);
    }, /null pointer|invalid memory/i);
    
    
    const ptr = wasm.trendflex_alloc(10);
    try {
        
        assert.throws(() => {
            wasm.trendflex_into(ptr, ptr, 10, 0);
        }, /Invalid period/);
        
        
        assert.throws(() => {
            wasm.trendflex_into(ptr, ptr, 10, 20);
        }, /Invalid period/);
    } finally {
        wasm.trendflex_free(ptr, 10);
    }
});

test('trendflex_batch_into', () => {
    const data = new Float64Array(testData.close.slice(0, 100));
    const period_start = 10;
    const period_end = 30;
    const period_step = 10;
    
    
    const expected_combos = 3;
    const total_size = expected_combos * data.length;
    
    
    const in_ptr = wasm.trendflex_alloc(data.length);
    const out_ptr = wasm.trendflex_alloc(total_size);
    
    try {
        
        const memory = wasm.__wasm.memory.buffer;
        const inView = new Float64Array(memory, in_ptr, data.length);
        inView.set(data);
        
        
        const n_combos = wasm.trendflex_batch_into(
            in_ptr, out_ptr, data.length,
            period_start, period_end, period_step
        );
        
        assert.strictEqual(n_combos, expected_combos);
        
        
        const memory2 = wasm.__wasm.memory.buffer;
        const outView = new Float64Array(memory2, out_ptr, total_size);
        
        
        const regularBatch = wasm.trendflex_batch(data, {
            period_range: [period_start, period_end, period_step]
        });
        
        assertArrayClose(
            Array.from(outView),
            regularBatch.values,
            1e-10,
            "Batch into vs regular batch mismatch"
        );
    } finally {
        wasm.trendflex_free(in_ptr, data.length);
        wasm.trendflex_free(out_ptr, total_size);
    }
});



test('trendflex_zero_copy_large_dataset', () => {
    const size = 100000;
    const data = new Float64Array(size);
    for (let i = 0; i < size; i++) {
        data[i] = Math.sin(i * 0.01) + Math.random() * 0.1;
    }
    
    const ptr = wasm.trendflex_alloc(size);
    assert(ptr !== 0, 'Failed to allocate large buffer');
    
    try {
        const memory = wasm.__wasm.memory.buffer;
        const memView = new Float64Array(memory, ptr, size);
        memView.set(data);
        
        wasm.trendflex_into(ptr, ptr, size, 20);
        
        
        const memory2 = wasm.__wasm.memory.buffer;
        const memView2 = new Float64Array(memory2, ptr, size);
        
        
        for (let i = 0; i < 20; i++) {
            assert(isNaN(memView2[i]), `Expected NaN at warmup index ${i}`);
        }
        
        
        for (let i = 20; i < Math.min(100, size); i++) {
            assert(!isNaN(memView2[i]), `Unexpected NaN at index ${i}`);
        }
    } finally {
        wasm.trendflex_free(ptr, size);
    }
});

test('trendflex_memory_management', () => {
    
    const sizes = [100, 1000, 10000, 50000];
    
    for (const size of sizes) {
        const ptr = wasm.trendflex_alloc(size);
        assert(ptr !== 0, `Failed to allocate ${size} elements`);
        
        
        const memory = wasm.__wasm.memory.buffer;
        const memView = new Float64Array(memory, ptr, size);
        for (let i = 0; i < Math.min(10, size); i++) {
            memView[i] = i * 1.5;
        }
        
        
        for (let i = 0; i < Math.min(10, size); i++) {
            assert.strictEqual(memView[i], i * 1.5, `Memory corruption at index ${i}`);
        }
        
        
        wasm.trendflex_free(ptr, size);
    }
});

test('trendflex_simd128_consistency', () => {
    
    
    const testCases = [
        { size: 10, period: 5 },
        { size: 100, period: 20 },
        { size: 1000, period: 30 },
        { size: 5000, period: 50 }
    ];
    
    for (const testCase of testCases) {
        const data = new Float64Array(testCase.size);
        for (let i = 0; i < testCase.size; i++) {
            data[i] = Math.sin(i * 0.1) + Math.cos(i * 0.05);
        }
        
        const result = wasm.trendflex_js(data, testCase.period);
        
        
        assert.strictEqual(result.length, data.length);
        
        
        for (let i = 0; i < testCase.period; i++) {
            assert(isNaN(result[i]), `Expected NaN at warmup index ${i} for size=${testCase.size}`);
        }
        
        
        let sumAfterWarmup = 0;
        let countAfterWarmup = 0;
        for (let i = testCase.period; i < result.length; i++) {
            assert(!isNaN(result[i]), `Unexpected NaN at index ${i} for size=${testCase.size}`);
            sumAfterWarmup += result[i];
            countAfterWarmup++;
        }
        
        
        const avgAfterWarmup = sumAfterWarmup / countAfterWarmup;
        assert(Math.abs(avgAfterWarmup) < 10, `Average value ${avgAfterWarmup} seems unreasonable`);
    }
});

test('trendflex_batch_metadata_comprehensive', () => {
    const close = new Float64Array(50);
    close.fill(100);
    
    const result = wasm.trendflex_batch(close, {
        period_range: [10, 20, 5]  
    });
    
    
    assert.strictEqual(result.combos.length, 3);
    assert.strictEqual(result.rows, 3);
    assert.strictEqual(result.cols, 50);
    
    
    const expectedPeriods = [10, 15, 20];
    for (let i = 0; i < expectedPeriods.length; i++) {
        assert.strictEqual(result.combos[i].period, expectedPeriods[i]);
    }
    
    
    for (let i = 0; i < result.rows; i++) {
        const rowStart = i * result.cols;
        const rowEnd = rowStart + result.cols;
        const rowData = result.values.slice(rowStart, rowEnd);
        
        
        const period = result.combos[i].period;
        for (let j = 0; j < period; j++) {
            assert(isNaN(rowData[j]), `Expected NaN at index ${j} for period ${period}`);
        }
        
        
        if (period < rowData.length) {
            assert(!isNaN(rowData[period]), `Expected valid value at index ${period} for period ${period}`);
        }
    }
});

test('trendflex_invalid_period_comprehensive', () => {
    const data = new Float64Array([1, 2, 3, 4, 5]);
    
    
    assert.throws(() => {
        wasm.trendflex_js(data, 0);
    }, /period = 0|ZeroTrendFlexPeriod/, 'Should reject period=0');
    
    
    assert.throws(() => {
        wasm.trendflex_js(data, 10);
    }, /period > data len|TrendFlexPeriodExceedsData/, 'Should reject period > length');
    
    
    assert.throws(() => {
        wasm.trendflex_js(data, data.length);
    }, /period > data len|TrendFlexPeriodExceedsData/, 'Should reject period = length');
});

test('trendflex_warmup_calculation_comprehensive', () => {
    
    const testPeriods = [5, 10, 20, 30];
    const close = testData.close.slice(0, 100); 
    
    for (const period of testPeriods) {
        if (period >= close.length) continue;
        
        const result = wasm.trendflex_js(close, period);
        
        
        const firstValid = 0;
        const warmup = firstValid + period;
        
        
        for (let i = 0; i < warmup; i++) {
            assert(isNaN(result[i]), `Expected NaN at index ${i} for period=${period}`);
        }
        
        
        if (warmup < result.length) {
            assert(!isNaN(result[warmup]), `Expected valid value at index ${warmup} for period=${period}`);
        }
    }
});
